### PR TITLE
[RLlib] POC: Run RLlib w/o Preprocessors setup.

### PR DESCRIPTION
**Pull request by sven1977**
This PR was motivated in preparation for soon allowing individual observation components to be addressed by the trajectory view API, for example to enable frame-stacking for individual observation components within a complex observation space (Tuple|Dict). Also, soon soft-deprecating RLlib's Preprocessor API should increase transparency for the users and allow batched, model-based preprocessing of observations. Observations will arrive in the model exactly as they are returned by the env.

This PR is a POC that works for tf and torch.

- New config setting for `preprocessor_pref`: None; Set to None for disabling Preprocessors altogether (not even use `NoPreprocessor` class anymore).
- A new example script was added (also to BUILD) to show the new setting in action with both tf and torch's ComplexInputModel (which already exists and handles complex, unflattened inputs).
- Most changes were necessary within the SimpleListCollector (SampleCollector) class allowing for arbitrarily nested data to be collected from the env and stored/retrieved.
- Works in unison with the recent SampleBatch enhancement (https://github.com/ray-project/ray/pull/17485) for allowing nested data within a SampleBatch.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/execution/replay_buffer.py [hunk 1]
```diff
@@ -402,7 +402,7 @@ def add_batch(self, batch: SampleBatchType) -> None:
                         # If SampleBatch has prio-replay weights, average
                         # over these to use as a weight for the entire
                         # sequence.
-                        if "weights" in time_slice:
+                        if "weights" in time_slice and time_slice["weights"]:
```
Comment (sven1977) [inline rllib/execution/replay_buffer.py:405]: Avoid np.mean over empty list (yields: NaN).

File: rllib/agents/dqn/dqn_tf_policy.py [hunk 2]
```diff
@@ -416,6 +416,8 @@ def postprocess_nstep_and_prio(policy: Policy,
                       batch[SampleBatch.REWARDS], batch[SampleBatch.NEXT_OBS],
                       batch[SampleBatch.DONES])
 
+    # Create dummy prio-weights (1.0) in case we don't have any in
```
Comment (michaelzhiluo) [inline rllib/agents/dqn/dqn_tf_policy.py:419]: Prob should be in diff PR

File: rllib/agents/trainer.py [hunk 3]
```diff
@@ -185,15 +185,17 @@
...
-    # env's action space before sending actions back to the env.
+    # (0.0 centered with small stddev; only affecting Box components).
+    # We will unsquash actions (and clip, just in case) to the bounds of
+    # the env's action space before sending actions back to the env.
     "normalize_actions": True,
     # If True, RLlib will clip actions according to the env's bounds
     # before sending them back to the env.
     # TODO: (sven) This option should be obsoleted and always be False.
     "clip_actions": False,
     # Whether to use "rllib" or "deepmind" preprocessors by default
```
Comment (michaelzhiluo) [inline rllib/agents/trainer.py:196]: Maybe describe what rllib or deepmind does in the comment.

File: rllib/evaluation/collectors/simple_list_collector.py [hunk 4]
```diff
@@ -80,19 +101,33 @@ def add_init_obs(self, episode_id: EpisodeID, agent_index: int,
...
-        self.buffers[SampleBatch.T].append(t)
+
+        # Append data to existing buffers.
+        flattened = tree.flatten(init_obs)
+        for i, sub_obs in enumerate(flattened):
+            self.buffers[SampleBatch.OBS][i].append(sub_obs)
+        self.buffers[SampleBatch.AGENT_INDEX][0].append(agent_index)
+        self.buffers[SampleBatch.ENV_ID][0].append(env_id)
+        self.buffers[SampleBatch.T][0].append(t)
+        self.buffers[SampleBatch.EPS_ID][0].append(self.episode_id)
```
Comment (michaelzhiluo) [inline rllib/evaluation/collectors/simple_list_collector.py:129]: If self.episode_id  and unroll_id is constant, why repeat adding the same data?

File: rllib/evaluation/collectors/simple_list_collector.py [hunk 5]
```diff
@@ -103,20 +138,40 @@ def add_action_reward_next_obs(self, values: Dict[str, TensorType]) -> \
...
         values[SampleBatch.OBS] = values[SampleBatch.NEXT_OBS]
         del values[SampleBatch.NEXT_OBS]
-        # Make sure EPS_ID stays the same for this agent. Usually, it should
-        # not be part of `values` anyways.
+
+        # Make sure EPS_ID/UNROLL_ID stay the same for this agent.
         if SampleBatch.EPS_ID in values:
             assert values[SampleBatch.EPS_ID] == self.episode_id
             del values[SampleBatch.EPS_ID]
+        self.buffers[SampleBatch.EPS_ID][0].append(self.episode_id)
```
Comment (michaelzhiluo) [inline rllib/evaluation/collectors/simple_list_collector.py:154]: same issue above

File: rllib/utils/tf_ops.py [hunk 6]
```diff
@@ -189,40 +208,42 @@ def call(*args, **kwargs):
...
-                                    dtype=v.dtype,
-                                    shape=shape,
-                                    name="kwarg_{}".format(k))
+                                shape = value.shape
+                            return tf1.placeholder(
+                                dtype=value.dtype,
+                                shape=shape,
+                                name=".".join([str(p) for p in path]),
+                            )
+
```
Comment (michaelzhiluo) [inline rllib/utils/tf_ops.py:231]: What does this do?

**Pull request closed without merging**